core/src/common.cpp

```diff
@@ -207,6 +218,119 @@ namespace irods::http
+		}};
+
+		return std::transform_reduce(
+			std::next(std::cbegin(_args)),
+			std::cend(_args),
+			encode_pair(*std::cbegin(_args)),
+			[](const auto& a, const auto& b) { return fmt::format("{}&{}", a, b); },
+			encode_pair);
+	}
+
+  auto safe_base64_encode(std::string_view _view) -> std::string {
+	constexpr auto char_per_byte_set{4};
+	constexpr auto byte_set_size{3};
+	auto max_size{char_per_byte_set * ((_view.size() + 2) / byte_set_size)};
+	auto max_size_plus_null_term{max_size + 1};
+		  
+	std::string encoded_data{};
+	encoded_data.resize(max_size);
+
+	auto res{irods::base64_encode(reinterpret_cast<const unsigned char*>(_view.data()),
+								  _view.size(),
+								  reinterpret_cast<unsigned char*>(encoded_data.data()),
+								  &max_size_plus_null_term)};
+	if (res) {
+	  log::debug("{}: Yikes!!!", __func__);
+	}
+	return encoded_data;
+  }
+  
+    auto hit_intro_endpoint(std::string _encoded_body) -> nlohmann::json
+	{
+		const auto introspection_endpoint{
+			irods::http::globals::oidc_endpoint_configuration().at("introspection_endpoint").get_ref<const std::string&>()};
+
+		const auto parsed_uri{boost::urls::parse_uri(introspection_endpoint)};
+
+		if (parsed_uri.has_error()) {
+			log::error(
+				"{}: Error trying to parse introspection_endpoint [{}]. Please check configuration.", __func__, introspection_endpoint);
+			return {{"error", "bad endpoint"}};
+		}
+
+		const auto url{*parsed_uri};
+		const auto port{get_port_from_url(url)};
+
+		// Addr
+		net::io_context io_ctx;
+		auto tcp_stream{irods::http::transport_factory(url.scheme_id(), io_ctx)};
+		tcp_stream->connect(url.host(), *port);
+
+		const auto format_bearer_token{[](std::string_view _client_id, std::string_view _client_secret){
+		  auto encode_me{fmt::format("{}:{}", encode(_client_id), encode(_client_secret))};
+		  return safe_base64_encode(encode_me);
+		}};
+
+		auto client_id{irods::http::globals::oidc_configuration().at("client_id").get_ref<const std::string&>()};
+		auto client_secret{irods::http::globals::oidc_configuration().at("client_secret").get_ref<const std::string&>()};
+		auto auth_string{fmt::format("Basic {}", format_bearer_token(client_id, client_secret))};
+	
+		// Build Request
+		constexpr auto version_number{11};
+		beast::http::request<beast::http::string_body> req{beast::http::verb::post, url.path(), version_number};
+		req.set(beast::http::field::host, url.host());
+		req.set(beast::http::field::user_agent, irods::http::version::server_name);
+		req.set(beast::http::field::content_type,
+		        "application/x-www-form-urlencoded"); // Possibly set a diff way?
+		req.set(beast::http::field::authorization, auth_string);
+		req.set(beast::http::field::accept, "application/json");
+
+		req.body() = std::move(_encoded_body);
+		req.prepare_payload();
+
+		// Send request & recieve response
+		auto res{tcp_stream->communicate(req)};
+		
+		log::debug("{}: Got the following resp back: [{}]", __func__, res.body());
```

Comment:
consider

"Received the following response:"

Reply:
Done.